EIPS/eip-7002.md

```diff
@@ -0,0 +1,372 @@
+
+To ensure that the withdrawal credentials (owned by both EOAs and smart contracts) can trustlessly control the destiny of the staked ETH, this specification enables exits triggerable by 0x01 withdrawal credentials.
+
+Note, 0x00 withdrawal credentials can be changed into 0x01 withdrawal credentials with a one-time signed message. Thus any functionality enabled for 0x01 credentials is defacto enabled for 0x00 credentials.
+
+## Specification
+
+### Constants
+
+| Name | Value | Comment |
+| - | - | - |
+|`FORK_TIMESTAMP` | *TBD* | Mainnet |
+
+### Configuration
+
+| Name | Value | Comment |
+| - | - | - |
+| `EXIT_PRECOMPILE_ADDRESS` | *TBD* | Where to call and store relevant details about exit mechanism |
+| `EXCESS_EXITS_STORAGE_SLOT` | 0 | |
+| `EXIT_COUNT_STORAGE_SLOT` | 1 | |
+| `EXIT_MESSAGE_QUEUE_HEAD_STORAGE_SLOT` | 2 | Pointer to head of the exit message queue |
+| `EXIT_MESSAGE_QUEUE_TAIL_STORAGE_SLOT` | 3 | Pointer to the tail of the exit message queue|
+| `EXIT_MESSAGE_QUEUE_STORAGE_OFFSET` | 4 | The start memory slot of the in-state exit message queue|
+| `MAX_EXITS_PER_BLOCK` | 16 | Maximum number of exits that can be dequeued into a block |
+| `TARGET_EXITS_PER_BLOCK` | 2 | |
+| `MIN_EXIT_FEE` | 1 | |
+| `EXIT_FEE_UPDATE_FRACTION` | 17 | | 
+| `EXCESS_RETURN_GAS_STIPEND` | 2300 | |
+
+### Execution layer
+
+#### Definitions
+
+* **`FORK_BLOCK`** -- the first block in a blockchain with the `timestamp` greater or equal to `FORK_TIMESTAMP`.
+
+#### Exit operation
+
+The new exit operation consists of the following fields:
+
+1. `source_address: Bytes20`
+2. `validator_index: uint64`
+
+RLP encoding of an exit **MUST** be computed as the following:
+
+```python
+rlp_encoded_deposit = RLP([
+    RLP(source_address),
+    RLP(validator_index),
+])
+```
+
+#### Exit precompile
+
+The precompile requires a single `8` byte big-endian `uint64` input, aliased to `validator_index`.
+
+`CALL`s to `EXIT_PRECOMPILE_ADDRESS` perform the following:
+
+* Ensure enough ETH was sent to cover the current exit fee (`check_exit_fee()`)
+* Increase exit count by 1 for the current block (`increment_exit_count()`)
+* Insert an exit into the queue for the source address and validator index (`insert_exit_to_queue()`)
+* Return any unspent ETH in excess of the exit fee with an `EXCESS_RETURN_GAS_STIPEND` gas stipend (`return_excess_payment()`)
+
+Specifically, the functionality is defined in pseudocode as the function `trigger_exit()`:
+
+```python
+###################
+# Public function #
+###################
+
+def trigger_exit(uint64 validator_index):
```

Comment:
Nit: type hints could be more pythonic `def trigger_exit(validator_index: uint64):`